psi4/driver/driver.py

```diff
@@ -873,6 +873,194 @@ def properties(*args, **kwargs):
         return core.variable('CURRENT ENERGY')
 
 
+
+def optimize_geometric(name, **kwargs):
+
+    import qcelemental as qcel
+    from qcelemental.util import which_import
+
+    if not which_import('geometric', return_bool=True):
+        raise ModuleNotFoundError('Python module geometric not found. Solve by installing it: `conda install -c conda-forge geometric` or `pip install geometric`')
+    import geometric
+
+    class Psi4NativeEngine(geometric.engine.Engine):
+        """
+        Internally run an energy and gradient calculation for geometric 
+        """
+        def __init__(self, p4_name, p4_mol, p4_return_wfn, **p4_kwargs):
+    
+            self.p4_name = p4_name
+            self.p4_mol = p4_mol
+            self.p4_return_wfn = p4_return_wfn
+            self.p4_kwargs = p4_kwargs
+    
+            molecule = geometric.molecule.Molecule()
+            molecule.elem = [p4_mol.symbol(i) for i in range(p4_mol.natom())]
+            molecule.xyzs = [p4_mol.geometry().np * qcel.constants.bohr2angstroms] 
+            molecule.build_bonds()
+                                 
+            super(Psi4NativeEngine, self).__init__(molecule)
+    
+        def calc(self, coords, dirname):
+            self.p4_mol.set_geometry(core.Matrix.from_array(coords.reshape(-1,3)))
+            self.p4_mol.update_geometry()
+            if self.p4_return_wfn:
+                g, wfn = gradient(self.p4_name, return_wfn=True, molecule=self.p4_mol, **self.p4_kwargs)
+                self.p4_wfn = wfn
+            else:
+                g = gradient(self.p4_name, return_wfn=False, molecule=self.p4_mol, **self.p4_kwargs)
+            e = core.variable('CURRENT ENERGY')
+            return {'energy': e, 'gradient': g.np.ravel()}
+
+    return_wfn = kwargs.pop('return_wfn', False)
+    return_history = kwargs.pop('return_history', False)
+
+    if return_history:
+        step_energies = []
+        step_gradients = []
+        step_coordinates = []
+
+    # Make sure the molecule the user provided is the active one
+    molecule = kwargs.get('molecule', core.get_active_molecule())
+
+    # Do not fix orientation or COM
+    molecule.fix_orientation(True)
+    molecule.fix_com(True)
+    molecule.update_geometry()
+
+    # Get geometric-specific options
+    geometric_opts = kwargs.get('geometric_opts', {})
+
+    core.print_out('\n')
+    core.print_out("\n  ==> GeomeTRIC Optimizer <==                                                                   ~\n")
+                                 
+    # Default to Psi4 maxiter unless overridden
+    if 'maxiter' not in geometric_opts:
```

Comment:
Do we need to worry about this being mixed / upper case at all?  I know @loriab sanitizes these things at various points, so maybe not

Reply:
Good point. kwargs are sanitized to lowercase keys [here](https://github.com/psi4/psi4/blob/master/psi4/driver/p4util/procutil.py#L47), but the geometric_opts dictionary won't be affected. Can fix with something like: `optimizer_keywords = {k.lower(): v for k, v in kwargs.get("optimizer_keywords", {}).items()}`

Is "geometric_opts" a name used in geometric? If not, may want something like "optimizer_keywords" to fall in line with qcsk and plan for future optimizers.

Reply:
Very helpful, thanks. `optimizer_keywords` does make more sense.